### PR TITLE
Render chat messages in fragments

### LINKED ISSUE #21: Render message templates with escaping
Right now, it looks like we're rendering any HTML content in chat messages passed through JSON from the Userscript. This is handy for rendering emote `` tags, but it opens the user up to rendering any HTML that wasn't already escaped by the streaming service before sending it down the socket/EventSource/etc. to the Userscript.

The `|safe` in this line is the culprit:
https://github.com/jaw-sh/stream-nexus/blob/a5a9e5b2c1fe0f43799473268f290952f56e67f1/templates/message.html#L23

The first solution that comes to mind is picking out any emote images and rendering them in their own template blocks. Something like `{{ msg_text_part_a }} {{ emote_img }} {{ msg_text_part_b}`. Splitting the messages into chunks like that can easily be handled by for loops in the message template.

### PR DESCRIPTION
**Pull request by ghost**
This renders messages reasonably safely by splitting them into fragments around emotes. I want to avoid all use of `|safe` because sanitation should never be assumed. This means I'll have to mess with username rendering because of role/sub tier badges as well.

Still a very early WIP. Making a draft PR for tracking and discussion.

fixes #21

### CODE REVIEW

**Comment by ghost** [comment]
Messages are now split up into `Fragment` objects, similar to nodes on an AST. Basically what every other parser does for an IR, except this is linear and not a tree.

**Comment by ghost** [comment]
Kick support is done. It works well in testing.

**Comment by jaw-sh** [comment]
I really, really wish you had talked to me before putting in all this work, because I don't know what this does and I'm going to have to review it to understand it.

**Comment by ghost** [comment]
It basically uses the existing emote regex to split the messages into chunks. Those chunks are stored in `Fragment` objects with an identifier for either `EMOTE` or `TEXT`. More fragment types can be added as needed. The `Fragments` work like nodes in trees used by most parsers, which tokenize and identify the node type (i.e. `EMOTE` or `TEXT`).

`EMOTE` fragments are required to have a valid URL as its associated value.

`TEXT` fragments are treated as plaintext, getting escaped by Askama.

We can use loops, like discussed in #21, to assemble the message body by rendering the chunks sequentially. `EMOTE` fragments get their url and id values put into an `` tag and rendered properly. `TEXT` fragments just get rendered as escaped message text. I moved some of the template stuff to Askama macros to better re-use the rendering templates.

Using fragments like this lets us be sure that we only render the HTML we expect to. The `|e` in the templates ensures all HTML stuff in plaintext gets escaped.

I apologize for not communicating enough before starting this. I will be sure to check in next time.

File: src/web/server.rs [hunk 1]
```diff
@@ -79,34 +106,21 @@ impl Handler<message::Content> for ChatServer {
...
-            replacement_string = replacement_string.replace(find, &format!("<{}>", key));
-            replacements.insert(key, value);
-        }
+        let mut chat_msg = msg.chat_message;
 
-        // Replace tokens with real replacements.
-        for (key, value) in replacements {
-            replacement_string = replacement_string.replace(&format!("<{}>", key), &value);
+        // Check if the userscript failed to generate a UUID.
+        if chat_msg.id.is_nil() {
```
Comment (ghost) [inline src/web/server.rs:112]: Setting a nil `UUID` in the default `Message` and checking if it's still present here helps with debugging if it somehow doesn't get overwritten during deserialization.

**Pull request closed without merging**